File: backend/app/providers/search/tavily.py
from collections.abc import Sequence
from typing import Any, Literal

from tavily import AsyncTavilyClient

from app.providers.search.base import ExtractedPage, MappedUrl, SearchResult, SearchTopic


class TavilySearchProvider:
    def __init__(self, api_key: str) -> None:
        self._client = AsyncTavilyClient(api_key=api_key)

    async def search(
        self,
        query: str,
        *,
        max_results: int = 5,
        search_depth: Literal["basic", "advanced"] = "basic",
        topic: SearchTopic = "general",
        include_domains: Sequence[str] | None = None,
        include_raw_content: bool | Literal["markdown", "text"] = False,
        time_range: Literal["day", "week", "month", "year"] | None = None,
    ) -> list[SearchResult]:
        kwargs: dict[str, Any] = {
            "query": query,
            "max_results": max_results,
            "search_depth": search_depth,
            "topic": topic,
            "include_raw_content": include_raw_content,
        }
        if include_domains:
            kwargs["include_domains"] = list(include_domains)
        if time_range:
            kwargs["time_range"] = time_range

        response = await self._client.search(**kwargs)
        return [
            SearchResult(
                url=r["url"],
                title=r.get("title", r["url"]),
                snippet=(r.get("content") or "")[:400],
                content=r.get("raw_content") or r.get("content"),
            )
            for r in response.get("results", [])
        ]

    async def extract(
        self,
        urls: Sequence[str],
        *,
        extract_depth: Literal["basic", "advanced"] = "basic",
    ) -> list[ExtractedPage]:
        if not urls:
            return []
        response = await self._client.extract(
            urls=list(urls),
            extract_depth=extract_depth,
            format="markdown",
        )
        return [
            ExtractedPage(
                url=r["url"],
                content=r.get("raw_content") or "",
                title=r.get("title"),
            )
            for r in response.get("results", [])
        ]

    async def map(
        self,
        url: str,
        *,
        max_depth: int = 2,
        limit: int = 30,
        categories: Sequence[str] | None = None,
    ) -> list[MappedUrl]:
        kwargs: dict[str, Any] = {
            "url": url,
            "max_depth": max_depth,
            "limit": limit,
        }
        if categories:
            kwargs["categories"] = list(categories)

        response = await self._client.map(**kwargs)
        items = response.get("results", [])
        out: list[MappedUrl] = []
        for r in items:
            if isinstance(r, str):
                out.append(MappedUrl(url=r))
            else:
                out.append(MappedUrl(url=r["url"], title=r.get("title")))
        return out

File: worker/src/tools/tavily.ts
import { env } from "@/config";

// Minimal Tavily REST client (search / extract / map) — mirrors the subset of
// app/providers/search/tavily.py the research tools use. No SDK dependency.

const BASE = "https://api.tavily.com";

async function post<T>(path: string, body: Record<string, unknown>): Promise<T> {
  const res = await fetch(`${BASE}${path}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${env.TAVILY_API_KEY}`,
    },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(60000),
  });
  if (!res.ok) throw new Error(`Tavily ${path} ${res.status}: ${await res.text()}`);
  return (await res.json()) as T;
}

export type SearchResult = {
  url: string;
  title: string;
  snippet: string;
  content: string | null;
};

type RawSearchHit = { url: string; title?: string; content?: string; raw_content?: string };

export async function tavilySearch(
  query: string,
  opts: {
    maxResults?: number;
    topic?: "general" | "news" | "finance";
    includeDomains?: string[];
    includeRawContent?: boolean | "markdown" | "text";
  } = {},
): Promise<SearchResult[]> {
  const body: Record<string, unknown> = {
    query,
    max_results: opts.maxResults ?? 4,
    search_depth: env.TAVILY_SEARCH_DEPTH,
    topic: opts.topic ?? "general",
    include_raw_content: opts.includeRawContent ?? false,
  };
  if (opts.includeDomains?.length) body.include_domains = opts.includeDomains;
  const data = await post<{ results?: RawSearchHit[] }>("/search", body);
  return (data.results ?? []).map((r) => ({
    url: r.url,
    title: r.title ?? r.url,
    snippet: (r.content ?? "").slice(0, 400),
    content: r.raw_content ?? r.content ?? null,
  }));
}

export type ExtractedPage = { url: string; content: string; title: string | null };

export async function tavilyExtract(urls: string[]): Promise<ExtractedPage[]> {
  if (urls.length === 0) return [];
  const data = await post<{ results?: { url: string; raw_content?: string; title?: string }[] }>(
    "/extract",
    { urls, extract_depth: "advanced", format: "markdown" },
  );
  return (data.results ?? []).map((r) => ({
    url: r.url,
    content: r.raw_content ?? "",
    title: r.title ?? null,
  }));
}

export type MappedUrl = { url: string; title: string | null };

export async function tavilyMap(
  url: string,
  opts: { maxDepth?: number; limit?: number; categories?: string[] } = {},
): Promise<MappedUrl[]> {
  const body: Record<string, unknown> = {
    url,
    max_depth: opts.maxDepth ?? 2,
    limit: opts.limit ?? 20,
  };
  if (opts.categories?.length) body.categories = opts.categories;
  const data = await post<{ results?: (string | { url: string; title?: string })[] }>("/map", body);
  return (data.results ?? []).map((r) =>
    typeof r === "string" ? { url: r, title: null } : { url: r.url, title: r.title ?? null },
  );
}
